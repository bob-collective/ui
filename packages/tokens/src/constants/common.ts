import { ChainId } from '@gobob/chains';
import { Bitcoin, ERC20Token } from '@gobob/currency';
import tokenList from 'tokenlist';

enum CommonToken {
  DAI = 'DAI',
  DLLR = 'DLLR',
  FBTC = 'FBTC',
  FRAX = 'FRAX',
  FXS = 'FXS',
  rETH = 'rETH',
  sFRAX = 'sFRAX',
  sfrxETH = 'sfrxETH',
  SOV = 'SOV',
  STONE = 'STONE',
  T = 'T',
  tBTC = 'tBTC',
  TRB = 'TRB',
  USDC = 'USDC',
  USDT = 'USDT',
  WBTC = 'WBTC',
  wstETH = 'wstETH'
}

const commonTokenSet = new Set(Object.values(CommonToken) as string[]);

function isCommonToken(symbol: string): symbol is CommonToken {
  return commonTokenSet.has(symbol);
}

const commonTokens = tokenList.tokens.reduce<Record<CommonToken, Record<ChainId, ERC20Token>>>(
  (acc, { symbol, chainId, address, decimals, name }) => {
    if (isCommonToken(symbol)) {
      if (acc[symbol] === undefined) acc[symbol] = {} as Record<ChainId, ERC20Token>;
      acc[symbol][chainId as ChainId] = new ERC20Token(chainId, address as `0x${string}`, decimals, symbol, name);
    }

    return acc;
  },
  {} as Record<CommonToken, Record<ChainId, ERC20Token>>
);

export const USDC = commonTokens[CommonToken.USDC];

export const WBTC = commonTokens[CommonToken.USDC];

export const TBTC = commonTokens[CommonToken.tBTC];

export const USDT_ETH = commonTokens[CommonToken.USDT][ChainId.ETHEREUM];

export const USDT = commonTokens[CommonToken.USDT];

export const BITCOIN = new Bitcoin(8, 'BTC', 'Bitcoin');
