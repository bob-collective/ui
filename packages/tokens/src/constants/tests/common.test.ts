import { describe, it, expect } from 'vitest';
import { ChainId } from '@gobob/chains';

import { commonTokens, CommonToken, WBTC, USDC, TBTC, USDT } from '../common';

describe('Common Tokens', () => {
  it('should have wBTC (WBTC) for chain ids 1 & 60808', () => {
    const tokenOnEthereum = commonTokens[CommonToken.WBTC]?.[ChainId.ETHEREUM];

    expect(tokenOnEthereum).toBeDefined();
    expect(tokenOnEthereum?.symbol).toBe(CommonToken.WBTC);

    const tokenOnBOB = commonTokens[CommonToken.WBTC]?.[ChainId.BOB];

    expect(tokenOnBOB).toBeDefined();
    expect(tokenOnBOB?.symbol).toBe(CommonToken.WBTC);
  });

  it('should have tBTC for chain ids 1 & 60808', () => {
    const tokenOnEthereum = commonTokens[CommonToken.tBTC]?.[ChainId.ETHEREUM];

    expect(tokenOnEthereum).toBeDefined();
    expect(tokenOnEthereum?.symbol).toBe(CommonToken.tBTC);

    const tokenOnBOB = commonTokens[CommonToken.tBTC]?.[ChainId.BOB];

    expect(tokenOnBOB).toBeDefined();
    expect(tokenOnBOB?.symbol).toBe(CommonToken.tBTC);
  });

  it('should have USDC for chain ids 1 & 60808', () => {
    const tokenOnEthereum = commonTokens[CommonToken.USDC]?.[ChainId.ETHEREUM];

    expect(tokenOnEthereum).toBeDefined();
    expect(tokenOnEthereum?.symbol).toBe(CommonToken.USDC);

    const tokenOnBOB = commonTokens[CommonToken.USDC]?.[ChainId.BOB];

    expect(tokenOnBOB).toBeDefined();
    expect(tokenOnBOB?.symbol).toBe(CommonToken.USDC);
  });

  it('should have USDT for chain ids 1 & 60808', () => {
    const tokenOnEthereum = commonTokens[CommonToken.USDT]?.[ChainId.ETHEREUM];

    expect(tokenOnEthereum).toBeDefined();
    expect(tokenOnEthereum?.symbol).toBe(CommonToken.USDT);

    const tokenOnBOB = commonTokens[CommonToken.USDT]?.[ChainId.BOB];

    expect(tokenOnBOB).toBeDefined();
    expect(tokenOnBOB?.symbol).toBe(CommonToken.USDT);
  });

  it('should have Dai (DAI) for chain ids 1 & 60808', () => {
    const tokenOnEthereum = commonTokens[CommonToken.DAI]?.[ChainId.ETHEREUM];

    expect(tokenOnEthereum).toBeDefined();
    expect(tokenOnEthereum?.symbol).toBe(CommonToken.DAI);

    const tokenOnBOB = commonTokens[CommonToken.DAI]?.[ChainId.BOB];

    expect(tokenOnBOB).toBeDefined();
    expect(tokenOnBOB?.symbol).toBe(CommonToken.DAI);
  });

  it('should match snapshot for WBTC', () => {
    expect(WBTC).toMatchSnapshot();
  });

  it('should match snapshot for USDC', () => {
    expect(USDC).toMatchSnapshot();
  });

  it('should match snapshot for TBTC', () => {
    expect(TBTC).toMatchSnapshot();
  });

  it('should match snapshot for USDT', () => {
    expect(USDT).toMatchSnapshot();
  });
});
