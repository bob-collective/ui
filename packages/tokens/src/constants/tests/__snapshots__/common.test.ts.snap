// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Common Tokens > should match snapshot for TBTC 1`] = `
{
  "1": ERC20Token {
    "address": "0x18084fbA666a33d37592fA2633fD49a74DD93a88",
    "chainId": 1,
    "decimals": 18,
    "isNative": false,
    "isToken": true,
    "name": "tBTC v2",
    "projectLink": undefined,
    "symbol": "tBTC",
  },
  "60808": ERC20Token {
    "address": "0xBBa2eF945D523C4e2608C9E1214C2Cc64D4fc2e2",
    "chainId": 60808,
    "decimals": 18,
    "isNative": false,
    "isToken": true,
    "name": "tBTC v2",
    "projectLink": undefined,
    "symbol": "tBTC",
  },
  "8453": ERC20Token {
    "address": "0x236aa50979D5f3De3Bd1Eeb40E81137F22ab794b",
    "chainId": 8453,
    "decimals": 18,
    "isNative": false,
    "isToken": true,
    "name": "Base tBTC v2",
    "projectLink": undefined,
    "symbol": "tBTC",
  },
}
`;

exports[`Common Tokens > should match snapshot for USDC 1`] = `
{
  "1": ERC20Token {
    "address": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    "chainId": 1,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USD Coin",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "10": ERC20Token {
    "address": "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
    "chainId": 10,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USD Coin",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "11155111": ERC20Token {
    "address": "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238",
    "chainId": 11155111,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USDC",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "11155420": ERC20Token {
    "address": "0x5fd84259d66Cd46123540766Be93DFE6D43130D7",
    "chainId": 11155420,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USDC",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "34443": ERC20Token {
    "address": "0xd988097fb8612cc24eeC14542bC03424c656005f",
    "chainId": 34443,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USD Coin",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "60808": ERC20Token {
    "address": "0xe75D0fB2C24A55cA1e3F96781a2bCC7bdba058F0",
    "chainId": 60808,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Bridged USDC (BOB)",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "808813": ERC20Token {
    "address": "0x2c1f73d37E7cdbA2792156139b0d35A765d595C5",
    "chainId": 808813,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Bridged USDC (BOB)",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "8453": ERC20Token {
    "address": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    "chainId": 8453,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USD Coin",
    "projectLink": undefined,
    "symbol": "USDC",
  },
  "84532": ERC20Token {
    "address": "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
    "chainId": 84532,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "USDC",
    "projectLink": undefined,
    "symbol": "USDC",
  },
}
`;

exports[`Common Tokens > should match snapshot for USDT 1`] = `
{
  "1": ERC20Token {
    "address": "0xdAC17F958D2ee523a2206206994597C13D831ec7",
    "chainId": 1,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Tether USD",
    "projectLink": undefined,
    "symbol": "USDT",
  },
  "10": ERC20Token {
    "address": "0x94b008aA00579c1307B0EF2c499aD98a8ce58e58",
    "chainId": 10,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Tether USD",
    "projectLink": undefined,
    "symbol": "USDT",
  },
  "1135": ERC20Token {
    "address": "0x05D032ac25d322df992303dCa074EE7392C117b9",
    "chainId": 1135,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Tether USD",
    "projectLink": undefined,
    "symbol": "USDT",
  },
  "34443": ERC20Token {
    "address": "0xf0F161fDA2712DB8b566946122a5af183995e2eD",
    "chainId": 34443,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Tether USD",
    "projectLink": undefined,
    "symbol": "USDT",
  },
  "424": ERC20Token {
    "address": "0x6535b3db9B908a2bbA29F83c168a0e661C3fAbf7",
    "chainId": 424,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Tether USD",
    "projectLink": undefined,
    "symbol": "USDT",
  },
  "60808": ERC20Token {
    "address": "0x05D032ac25d322df992303dCa074EE7392C117b9",
    "chainId": 60808,
    "decimals": 6,
    "isNative": false,
    "isToken": true,
    "name": "Tether USD",
    "projectLink": undefined,
    "symbol": "USDT",
  },
}
`;

exports[`Common Tokens > should match snapshot for WBTC 1`] = `
{
  "1": ERC20Token {
    "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "chainId": 1,
    "decimals": 8,
    "isNative": false,
    "isToken": true,
    "name": "Wrapped BTC",
    "projectLink": undefined,
    "symbol": "WBTC",
  },
  "10": ERC20Token {
    "address": "0x68f180fcCe6836688e9084f035309E29Bf0A2095",
    "chainId": 10,
    "decimals": 8,
    "isNative": false,
    "isToken": true,
    "name": "Wrapped BTC",
    "projectLink": undefined,
    "symbol": "WBTC",
  },
  "34443": ERC20Token {
    "address": "0xcDd475325D6F564d27247D1DddBb0DAc6fA0a5CF",
    "chainId": 34443,
    "decimals": 8,
    "isNative": false,
    "isToken": true,
    "name": "Wrapped BTC",
    "projectLink": undefined,
    "symbol": "WBTC",
  },
  "60808": ERC20Token {
    "address": "0x03C7054BCB39f7b2e5B2c7AcB37583e32D70Cfa3",
    "chainId": 60808,
    "decimals": 8,
    "isNative": false,
    "isToken": true,
    "name": "Wrapped BTC",
    "projectLink": undefined,
    "symbol": "WBTC",
  },
}
`;
